File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/view/fragments/AddToDeckFragment.java
package com.dbottillo.mtgsearchfree.view.fragments;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.DialogFragment;
import android.text.InputFilter;
import android.text.Spanned;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;

import com.dbottillo.mtgsearchfree.MTGApp;
import com.dbottillo.mtgsearchfree.R;
import com.dbottillo.mtgsearchfree.model.Deck;
import com.dbottillo.mtgsearchfree.model.MTGCard;
import com.dbottillo.mtgsearchfree.presenter.DecksPresenter;
import com.dbottillo.mtgsearchfree.util.TrackingManager;
import com.dbottillo.mtgsearchfree.view.DecksView;

import java.util.List;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class AddToDeckFragment extends BasicFragment implements DecksView {

    @Bind(R.id.choose_deck)
    Spinner chooseDeck;
    @Bind(R.id.choose_quantity)
    Spinner chooseQuantity;
    @Bind(R.id.add_to_deck_sideboard)
    CheckBox sideboard;

    String[] decksChoose;
    String[] quantityChoose;

    @Bind(R.id.new_deck_name_input_layout)
    TextInputLayout cardNameInputLayout;
    @Bind(R.id.new_deck_name)
    EditText deckName;
    @Bind(R.id.new_deck_quantity_input_layout)
    TextInputLayout cardQuantityInputLayout;
    @Bind(R.id.new_deck_quantity)
    EditText cardQuantity;

    List<Deck> decks;
    MTGCard card;

    @Inject
    DecksPresenter decksPresenter;

    public static DialogFragment newInstance(MTGCard card) {
        AddToDeckFragment instance = new AddToDeckFragment();
        Bundle args = new Bundle();
        args.putParcelable("card", card);
        instance.setArguments(args);
        return instance;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_add_to_deck, container, false);
        ButterKnife.bind(this, v);
        return v;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        card = getArguments().getParcelable("card");
        cardQuantity.setFilters(new InputFilter[]{new InputFilterMinMax(1, 30)});

        setupQuantitySpinner();

        MTGApp.dataGraph.inject(this);
        decksPresenter.init(this);
        decksPresenter.loadDecks();
    }

    private void setupQuantitySpinner() {
        quantityChoose = new String[]{getString(R.string.deck_choose_quantity), "1", "2", "3", "4", getString(R.string.deck_specify)};
        final ArrayAdapter<CharSequence> adapter = new ArrayAdapter<CharSequence>(getActivity(), R.layout.add_to_deck_spinner_item, quantityChoose);
        adapter.setDropDownViewResource(R.layout.add_to_deck_dropdown_item);
        chooseQuantity.setAdapter(adapter);
        chooseQuantity.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                if (position == 5) {
                    chooseQuantity.setVisibility(View.GONE);
                    cardQuantityInputLayout.setVisibility(View.VISIBLE);
                    cardQuantity.requestFocus();
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Dialog dialog = super.onCreateDialog(savedInstanceState);
        // request a window without the title
        dialog.getWindow().requestFeature(Window.FEATURE_NO_TITLE);
        return dialog;
    }

    @Override
    public String getPageTrack() {
        return "/add_to_deck";
    }

    private void setupDecksSpinner(final List<Deck> decks) {
        this.decks = decks;
        decksChoose = new String[decks.size() + 2];
        decksChoose[0] = getString(R.string.deck_choose);
        int i = 1;
        for (Deck deck : decks) {
            decksChoose[i] = deck.getName();
            i++;
        }
        decksChoose[decks.size() + 1] = getString(R.string.deck_new);
        final ArrayAdapter<CharSequence> adapter = new ArrayAdapter<CharSequence>(getActivity(), R.layout.add_to_deck_spinner_item, decksChoose);
        adapter.setDropDownViewResource(R.layout.add_to_deck_dropdown_item);
        chooseDeck.setAdapter(adapter);
        chooseDeck.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                if (position == decks.size() + 1) {
                    chooseDeck.setVisibility(View.GONE);
                    cardNameInputLayout.setVisibility(View.VISIBLE);
                    deckName.requestFocus();
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });
    }

    @OnClick(R.id.add_to_deck_save)
    public void addToDeck(View view) {
        int quantity = -1;
        if (chooseQuantity.getVisibility() == View.VISIBLE && chooseQuantity.getSelectedItemPosition() > 0) {
            quantity = Integer.parseInt(quantityChoose[chooseQuantity.getSelectedItemPosition()]);
        }
        if (chooseQuantity.getVisibility() == View.GONE && cardQuantity.getText().length() > 0) {
            quantity = Integer.parseInt(cardQuantity.getText().toString());
        }
        if (quantity > -1) {
            if (chooseDeck.getVisibility() == View.VISIBLE && chooseDeck.getSelectedItemPosition() > 0) {
                Deck deck = decks.get(chooseDeck.getSelectedItemPosition() - 1);
                boolean side = sideboard.isChecked();
                saveCard(quantity, deck, side);
                dismiss();
            }
            if (chooseDeck.getVisibility() == View.GONE && deckName.getText().length() > 0) {
                boolean side = sideboard.isChecked();
                saveCard(quantity, deckName.getText().toString(), side);
                dismiss();
            }
        }
    }

    private void saveCard(final int quantity, final Deck deck, final boolean side) {
        card.setSideboard(side);
        decksPresenter.addCardToDeck(deck, card, quantity);
        TrackingManager.trackAddCardToDeck(quantity + " - existing");
    }

    private void saveCard(final int quantity, final String deck, final boolean side) {
        card.setSideboard(side);
        decksPresenter.addCardToDeck(deck, card, quantity);
        TrackingManager.trackNewDeck(deck);
        TrackingManager.trackAddCardToDeck(quantity + " - existing");
    }

    @Override
    public void decksLoaded(List<Deck> decks) {
        setupDecksSpinner(decks);
    }

    @Override
    public void deckLoaded(List<MTGCard> cards) {
    }

    @Override
    public void showError(String message) {
        Toast.makeText(getContext(), message, Toast.LENGTH_SHORT).show();
    }

    public static class InputFilterMinMax implements InputFilter {

        private int min, max;

        public InputFilterMinMax(int min, int max) {
            this.min = min;
            this.max = max;
        }

        @Override
        public CharSequence filter(CharSequence source, int start, int end, Spanned dest, int dstart, int dend) {
            try {
                int input = Integer.parseInt(dest.toString() + source.toString());
                if (isInRange(min, max, input)) {
                    return null;
                }
            } catch (NumberFormatException ignored) {
                return "";
            }
            return "";
        }

        private boolean isInRange(int a, int b, int c) {
            return b > a ? c >= a && c <= b : c >= b && c <= a;
        }
    }

}

File: MTGSearch/src/main/kotlin/com/dbottillo/mtgsearchfree/ui/decks/addToDeck/AddToDeckFragment.kt
package com.dbottillo.mtgsearchfree.ui.decks.addToDeck

import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.support.design.widget.TextInputLayout
import android.support.v4.app.DialogFragment
import android.text.InputFilter
import android.text.Spanned
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.widget.*
import com.dbottillo.mtgsearchfree.R
import com.dbottillo.mtgsearchfree.model.Deck
import com.dbottillo.mtgsearchfree.model.MTGCard
import com.dbottillo.mtgsearchfree.ui.BasicFragment
import com.dbottillo.mtgsearchfree.util.LOG
import com.dbottillo.mtgsearchfree.util.TrackingManager
import javax.inject.Inject
import dagger.android.support.AndroidSupportInjection



class AddToDeckFragment : BasicFragment(), AddToDeckView {

    lateinit var chooseDeck: Spinner
    lateinit var chooseQuantity: Spinner
    lateinit var sideboard: CheckBox
    lateinit var cardNameInputLayout: TextInputLayout
    lateinit var deckName: EditText
    lateinit var cardQuantityInputLayout: TextInputLayout
    lateinit var cardQuantity: EditText
    lateinit var title: TextView

    internal var decksChoose: MutableList<String> = mutableListOf()
    internal var quantityChoose: Array<String> = arrayOf()

    internal var decks: List<Deck> = mutableListOf()

    @Inject
    lateinit var presenter: AddToDeckPresenter

    override fun onAttach(context: Context) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_add_to_deck, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        title = view.findViewById(R.id.add_card_title)
        chooseDeck = view.findViewById(R.id.choose_deck)
        chooseQuantity = view.findViewById(R.id.choose_quantity)
        sideboard = view.findViewById(R.id.add_to_deck_sideboard)
        cardNameInputLayout = view.findViewById(R.id.new_deck_name_input_layout)
        cardQuantityInputLayout = view.findViewById(R.id.new_deck_quantity_input_layout)
        deckName = view.findViewById(R.id.new_deck_name)
        cardQuantity = view.findViewById(R.id.new_deck_quantity)
        view.findViewById<View>(R.id.add_to_deck_save).setOnClickListener { addToDeck() }

        cardQuantity.filters = arrayOf<InputFilter>(InputFilterMinMax(1, 30))

        setupQuantitySpinner()

        presenter.init(this, arguments)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        presenter.onDestroyView()
    }

    private fun setupQuantitySpinner() {
        LOG.d()
        quantityChoose = arrayOf(getString(R.string.deck_choose_quantity), "1", "2", "3", "4", getString(R.string.deck_specify))
        val adapter = ArrayAdapter<CharSequence>(activity, R.layout.add_to_deck_spinner_item, quantityChoose)
        adapter.setDropDownViewResource(R.layout.add_to_deck_dropdown_item)
        chooseQuantity.adapter = adapter
        chooseQuantity.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View?, position: Int, id: Long) {
                LOG.d()
                if (position == 5) {
                    chooseQuantity.visibility = View.GONE
                    cardQuantityInputLayout.visibility = View.VISIBLE
                    cardQuantity.requestFocus()
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>) {

            }
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        LOG.d()
        val dialog = super.onCreateDialog(savedInstanceState)
        // request a window without the title
        dialog.window.requestFeature(Window.FEATURE_NO_TITLE)
        return dialog
    }

    override fun getPageTrack(): String {
        return "/add_to_deck"
    }

    private fun setupDecksSpinner(decks: List<Deck>) {
        LOG.d()
        this.decks = decks
        decksChoose.clear()
        decksChoose.add(getString(R.string.deck_choose))
        decks.forEach { decksChoose.add(it.name) }
        decksChoose.add(getString(R.string.deck_new))
        val adapter = ArrayAdapter<CharSequence>(activity, R.layout.add_to_deck_spinner_item, decksChoose.toTypedArray())
        adapter.setDropDownViewResource(R.layout.add_to_deck_dropdown_item)
        chooseDeck.adapter = adapter
        chooseDeck.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                if (position == decks.size + 1) {
                    chooseDeck.visibility = View.GONE
                    cardNameInputLayout.visibility = View.VISIBLE
                    deckName.requestFocus()
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>) {

            }
        }
    }

    internal fun addToDeck() {
        LOG.d()
        var quantity = -1
        if (chooseQuantity.visibility == View.VISIBLE && chooseQuantity.selectedItemPosition > 0) {
            quantity = Integer.parseInt(quantityChoose[chooseQuantity.selectedItemPosition])
        }
        if (chooseQuantity.visibility == View.GONE && cardQuantity.text.isNotEmpty()) {
            quantity = Integer.parseInt(cardQuantity.text.toString())
        }
        if (quantity > -1) {
            if (chooseDeck.visibility == View.VISIBLE && chooseDeck.selectedItemPosition > 0) {
                val deck = decks[chooseDeck.selectedItemPosition - 1]
                val side = sideboard.isChecked
                saveCard(quantity, deck, side)
                dismiss()
            }
            if (chooseDeck.visibility == View.GONE && deckName.text.isNotEmpty()) {
                val side = sideboard.isChecked
                saveCard(quantity, deckName.text.toString(), side)
                dismiss()
            }
        }
    }

    private fun saveCard(quantity: Int, deck: Deck, side: Boolean) {
        LOG.d()
        presenter.addCardToDeck(deck, quantity, side)
        TrackingManager.trackAddCardToDeck(quantity.toString() + " - existing")
    }

    private fun saveCard(quantity: Int, deck: String, side: Boolean) {
        LOG.d()
        presenter.addCardToDeck(deck, quantity, side)
        TrackingManager.trackNewDeck(deck)
        TrackingManager.trackAddCardToDeck(quantity.toString() + " - existing")
    }

    override fun showError(message: String) {
        Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
    }

    override fun setCardTitle(cardName: String) {
        title.text = getString(R.string.add_to_deck_title, cardName)
    }

    override fun decksLoaded(decks: List<Deck>) {
        LOG.d()
        setupDecksSpinner(decks)
    }

    class InputFilterMinMax(private val min: Int, private val max: Int) : InputFilter {

        override fun filter(source: CharSequence, start: Int, end: Int, dest: Spanned, dstart: Int, dend: Int): CharSequence? {
            try {
                val input = Integer.parseInt(dest.toString() + source.toString())
                if (isInRange(min, max, input)) {
                    return null
                }
            } catch (ignored: NumberFormatException) {
                return ""
            }

            return ""
        }

        private fun isInRange(a: Int, b: Int, c: Int): Boolean {
            return if (b > a) c in a..b else c in b..a
        }
    }

    companion object {

        fun newInstance(card: MTGCard): DialogFragment {
            val instance = AddToDeckFragment()
            val args = Bundle()
            args.putInt("card", card.multiVerseId)
            instance.arguments = args
            return instance
        }
    }

    override fun getTitle(): String {
        return "/add_to_deck"
    }
}
